pkg/bpf/collection.go

```diff
@@ -41,6 +48,138 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			log.Debugf("section '%s' / prog '%s', not a tailcall, considering an entrypoint", prog.SectionName, prog.Name)
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		const (
+			cilium_calls_map = 2
+		)
+
+		// Any program that does follow the x/y naming convention but not part
+		// of the cilium_calls map is also considered an entrypoint.
+		if id != cilium_calls_map {
+			log.Debugf("section '%s' / prog '%s', not a cilium_calls, considering an entrypoint", prog.SectionName, prog.Name)
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		if tailcalls[slot] != nil {
+			return fmt.Errorf("duplicate tail call index %d", slot)
+		}
+
+		tailcalls[slot] = &TailCall{
+			spec: prog,
+		}
+	}
+
+	// Discover all tailcalls that are reachable from the given program.
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) error {
+		log.Debugf("visiting section '%s' / prog '%s'", prog.SectionName, prog.Name)
+
+		realOff := 2
+		for i := 3; i < len(prog.Instructions); i++ {
+			// The `tail_call_static` C function is always used to call tail calls when
+			// the map index is known at compile time.
+			// Due to inline ASM this generates the following instructions:
+			//   Mov R1, Rx
+			//   Mov R2, <map>
+			//   Mov R3, <index>
+			//   call tail_call
+
+			// Find the tail call instruction.
+			inst := prog.Instructions[i]
+			realOff++
+			if inst.OpCode.IsDWordLoad() {
+				realOff++
+			}
+			if !inst.IsBuiltinCall() || inst.Constant != int64(asm.FnTailCall) {
+				continue
+			}
+
+			// Check that the previous instruction is a mov of the tail call index.
+			movIdx := prog.Instructions[i-1]
+			if movIdx.OpCode.ALUOp() != asm.Mov || movIdx.Dst != asm.R3 {
+				log.Debugf("found tail call @ %d, R3 mismatch, skipping", realOff)
+				continue
+			}
+
+			// Check that the instruction before that is the load of the tail call map.
+			movR2 := prog.Instructions[i-2]
+			if movR2.OpCode != asm.LoadImmOp(asm.DWord) || movR2.Src != asm.PseudoMapFD {
+				log.Debugf("found tail call @ %d, R2 mismatch, skipping", realOff)
+				continue
+			}
+
+			ref := movR2.Reference()
+
+			// Ignore static tail calls made to maps that are not the calls map
+			if !strings.Contains(ref, callsmap.MapName) || strings.Contains(ref, callsmap.CustomCallsMapName) {
+				log.Debugf("found tail call @ %d, reference '%s', not a calls map, skipping", realOff, ref)
+				continue
+			}
+
+			tc := tailcalls[uint32(movIdx.Constant)]
+			if tc == nil {
+				return fmt.Errorf("tail call to unknown index '%d' @ %d, potential missed tailcall", movIdx.Constant, realOff)
```

Comment:
Not sure if insn index is valuable to the reader, I'd remove it.

```suggestion
				return fmt.Errorf("missed tail call to program at cilium_calls slot %d", movIdx.Constant)
```